trieste/models/gpflow/models.py

```diff
@@ -1647,3 +1647,367 @@ def covariance_with_top_fidelity(self, query_points: TensorType) -> TensorType:
+        query_points_fidelity_col = tf.repeat(
+            query_points_fidelity_col, self.monte_carlo_random_numbers.shape[0], axis=-1
+        )[
+            ..., :, None, :
+        ]  # [N, 1, S]
+
+        # Predict for all fidelities but stop updating once we have
+        # reached desired fidelity for each query point
+        for fidelity in range(1, self.num_fidelities):
+
+            # sample_mean [..., N, 1, S]
+            # sample_var [..., N, 1, S]
+            (
+                next_fidelity_sample_mean,
+                next_fidelity_sample_var,
+            ) = self._propogate_samples_through_level(
+                query_points_wo_fidelity, fidelity, sample_mean, sample_var
+            )
+
+            mask = query_points_fidelity_col >= fidelity  # [..., N, 1, S]
+
+            sample_mean = tf.where(mask, next_fidelity_sample_mean, sample_mean)  # [..., N, 1, S]
+            sample_var = tf.where(mask, next_fidelity_sample_var, sample_var)  # [..., N, 1, S]
+
+        return sample_mean, sample_var
+
+    def _propogate_samples_through_level(
+        self,
+        query_point: TensorType,
+        fidelity: int,
+        sample_mean: TensorType,
+        sample_var: TensorType,
+    ) -> tuple[TensorType, TensorType]:
+        """
+        Propogate samples through a given fidelity.
+
+        This takes a set of query points without a fidelity column and calculates samples
+        at the given fidelity, using the sample means and variances from the previous fidelity.
+
+        :param query_points: The query points to sample at, with no fidelity column,
+            with shape[..., N, D]
+        :param fidelity: The fidelity to propogate the samples through
+        :param sample_mean: Samples of the posterior mean at the previous fidelity,
+            with shape [..., N, 1, S]
+        :param sample_var: Samples of the posterior variance at the previous fidelity,
+            with shape [..., N, 1, S]
+        :return: sample_mean: Samples of the posterior mean at the given fidelity,
+            of shape [..., N, 1, S]
+            and sample_var: Samples of the posterior variance at the given fidelity,
+            of shape [..., N, 1, S]
+        """
+        B, N, D = query_point.shape[:-2], query_point.shape[-2], query_point.shape[-1]
+        S = sample_mean.shape[-1]
+        # Repeat random numbers for each query point and add middle dimension
+        # for concatentation with query points This also means that it has the same
+        # shape as sample_var and sample_mean, so there's no broadcasting required
+        # Note: at the moment we use the same monte carlo values for every value in the batch dim
+        reshaped_random_numbers = tf.repeat(
+            tf.transpose(self.monte_carlo_random_numbers), N, axis=-2
+        )[
+            ..., None, :
+        ]  # [..., N, 1, S]
+        samples = reshaped_random_numbers * tf.sqrt(sample_var) + sample_mean  # [..., N, 1, S]
```

Comment:
actually I think that if `sample_var` and `sample_mean` are of shape `[..., N, 1, 1]` this will just broadcast? so you don't need the repeats above. But looks like they are useful for the mask?

Reply:
Yeah, I did everything explicitly and avoided broadcasting to try and keep better track of everything. But yes, I think for mask it's useful to have everything the same shape or it can get confusing (for me at least!)

Reply:
if this doesn't incur any extra cost (or memory print) then it's not worth bothering I guess?